### PR TITLE
Add kernel linting via `@parcels.validate_kernel`

### LINKED ISSUE #2109: Runtime warnings for kernels (syntax upgrading, and correctness)
_Part of #2106_

With the JIT code conversion gone in v4-dev of Parcels, there are some warnings that were implemented in `codegenerator.py` regarding correctness:

- ["Don't change the location of a particle directly in a Kernel. Use particle_dlon, particle_dlat, etc."](https://github.com/OceanParcels/Parcels/blob/88cd174af40d8446ac2ab8054687e9b381da9078/parcels/compilation/codegenerator.py#L317-L322)
- ["Be careful when sampling particle.(xi/yi/zi), as this is updated in the kernel loop. Best to place the sampling statement before advection."](https://github.com/OceanParcels/Parcels/blob/88cd174af40d8446ac2ab8054687e9b381da9078/parcels/compilation/codegenerator.py#L181-L190)

That aren't implemented anymore. AFAIK (please confirm @erikvansebille) these warnings only came up during JIT mode, and never during Scipy mode.

Additionally, with the kernel syntax changes we can expect to land in v4 (see parent issue), having warnings in the following format would be helpful:

- `` `particle.delete()` is no longer valid syntax in Parcels kernels - it looks like you're using a kernel written for parcels<4 . HINT: Use `particle.status = StatusCode.Delete` instead. See our migration guide at https://docs.oceanparcels.org/... for more info.``

This would also allow us to identify kernels that are written for older versions of Parcels (not just for v3->v4, but also for older versions if we want), and giving users helpful error messages with instructions for upgrading their kernels.

Suggestion:
- Before compiling kernels with numba, we have a kernel linting step.
  - This will inspect the source code and:
    - Error out informatively if the kernel is incompatible with the new version of Parcels
    - Provide a warning if correctness is a concern
  - Importantly, the users code will not be altered under the hood.

### PR DESCRIPTION
**Pull request by VeckoTheGecko**
### Description

This PR adds linting for Parcels kernels to help with correctness and for migration to v4.

The linting is broken up into errors and warnings:
- Errors are guaranteed to be incorrect
  - i.e., using an incorrect function signature for your kernels, calling `particles.delete()`
- Warnings are either likely to be incorrect, or give inaccurate results
  - using `particle_dlon` - a user might intentionally define this variable in their v4 kernel. There's no way to distinguish between that usage and someone upgrading their v3 kernel.
  - sampling `ei`, they should be warned about where they're sampling it

This linting is exposed via a decorator for users, and the docstring for `validate_kernel` shows which rules and looked for. I looked through the migration guide etc. and thought these rules were good, let me know @erikvansebille if you think additional rules are needed.

```python
@parcels.validate_kernel
def my_kernel(particle, fieldset, time): ...
```

providing an opt-in endpoint for raising these errors and warnings before the execution phase of a simulation

Internally in the execute loop these kernels are validated again. Providing the decorator is more for providing a good user experience for upgrading kernels.

#### Example kernel upgrade

The following kernel is taken from PlasticParcels

```python
import parcels
import math

@parcels.validate_kernel
def StokesDrift(particle, fieldset, time):
    # Sample the peak wave period
    T_p = fieldset.wave_Tp[time, particle.depth, particle.lat, particle.lon]

    # Compute the local bathymetry / water depth with a margin of error
    local_bathymetry = 0.99*fieldset.bathymetry[time, particle.depth, particle.lat, particle.lon]

    # Only compute displacements if the peak wave period is large enough and the particle is in the water
    if T_p > 1E-14 and particle.depth < local_bathymetry:
        # Sample the U / V components of Stokes drift
        stokes_U = fieldset.Stokes_U[time, particle.depth, particle.lat, particle.lon]
        stokes_V = fieldset.Stokes_V[time, particle.depth, particle.lat, particle.lon]

        # Peak wave frequency
        omega_p = 2. * math.pi / T_p

        # Peak wave number
        k_p = (omega_p ** 2) / fieldset.G

        # Repeated inner term of Eq. (19) - note depth is negative in this formulation, but model depths are positive by convention
        kp_z_2 = 2. * k_p * particle.depth

        # Decay factor in Eq. (19) -- Where beta=1 for the Phillips spectrum
        decay = math.exp(-kp_z_2) - math.sqrt(math.pi * kp_z_2) * math.erfc(math.sqrt(kp_z_2))

        # Apply Eq. (19) and compute particle displacement
        particle_dlon += stokes_U * decay * particle.dt  # noqa
        particle_dlat += stokes_V * decay * particle.dt  # noqa
```

which gives

```
/Users/Hodgs004/coding/repos/parcels/t.py:1: UserWarning: This is an alpha version of Parcels v4. The API is not stable and may change without deprecation warnings.
  import parcels
/Users/Hodgs004/coding/repos/parcels/t.py:4: KernelWarning: Kernel `StokesDrift` has 1 validation error(s):
  - Kernel signature must be `def StokesDrift(particles, fieldset)`, but got `def StokesDrift(particle, fieldset, time)`. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.

Kernel `StokesDrift` has 16 warning(s):
  - Line 4: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 4: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 4: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 10: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 22: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 28: `particle_dlon` is no longer valid in Parcels v4. Use `particles.dx` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 29: `particle_dlat` is no longer valid in Parcels v4. Use `particles.dy` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  @parcels.validate_kernel
Traceback (most recent call last):
  File "/Users/Hodgs004/coding/repos/parcels/t.py", line 4, in <module>
    @parcels.validate_kernel
     ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/Hodgs004/coding/repos/parcels/src/parcels/_core/utils/kernel_linting.py", line 232, in validate_kernel
    raise KernelValidationError(report)
parcels._core.utils.kernel_linting.KernelValidationError: Kernel `StokesDrift` has 1 validation error(s):
  - Kernel signature must be `def StokesDrift(particles, fieldset)`, but got `def StokesDrift(particle, fieldset, time)`. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.

Kernel `StokesDrift` has 16 warning(s):
  - Line 4: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 4: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 4: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 7: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 10: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 12: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.lat` is deprecated. Use `particles.y` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 13: `particle.lon` is deprecated. Use `particles.x` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 22: `particle.depth` is deprecated. Use `particles.z` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 28: `particle_dlon` is no longer valid in Parcels v4. Use `particles.dx` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
  - Line 29: `particle_dlat` is no longer valid in Parcels v4. Use `particles.dy` instead. See https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html for more info.
```

Note that there is a doubling up - the "report" is printed once for the warnings and once for the errors. I think that this is fine, and better than splitting them up (as warnings are likely errors as well, its nice to deal with them at the same time).

### Checklist

- [x] Closes #2109
- [x] Tests added
- [x] This PR targets the correct branch (`main` for normal development, `v3-support` for v3 support)

### AI Disclosure

- [x] This PR contains AI-generated content.
  - [x] I have tested any AI-generated content in my PR.
  - [x] I take responsibility for any AI-generated content in my PR.
  - Describe how you used it (e.g., by pasting your prompt): For creating the AST parsers and for creating the test suite

### CODE REVIEW

**Comment by VeckoTheGecko** [comment]
Still need to update the tests (will do next week), otherwise ready for review

We need to decide where in docs to raise this (likely in quickstart and in migration guide)

File: src/parcels/_core/utils/kernel_linting.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+import textwrap
+import types
+import warnings
+from collections.abc import Callable, Sequence
+from dataclasses import dataclass, field
+from typing import TypeVar
+
+from parcels._core.warnings import KernelWarning
+
+_V4_MIGRATION_GUIDE: str = "https://docs.oceanparcels.org/en/latest/user_guide/v4-migration.html"
```
Comment (erikvansebille) [inline src/parcels/_core/utils/kernel_linting.py]: Use the new URL ;-)
```suggestion
_V4_MIGRATION_GUIDE: str = "https://docs.parcels-code.org/en/latest/user_guide/v4-migration.html"
```

**Review by erikvansebille — COMMENTED** [review]
Looks great! One suggested change below - and I added references to this at four places in the documentation (5de90188f5b60434cb102253235cd749f0b11e84)

File: docs/user_guide/getting_started/explanation_concepts.md [hunk 2]
```diff
@@ -116,13 +116,16 @@ where $\mathbf{v}(\mathbf{x},t) = (u(\mathbf{x},t), v(\mathbf{x},t))$ describes
...
 ```python
+@parcels.validate_kernel
 def AdvectionEE(particles, fieldset):
     """Advection of particles using Explicit Euler (aka Euler Forward) integration."""
     (u1, v1) = fieldset.UV[particles]
     particles.dx += u1 * particles.dt
     particles.dy += v1 * particles.dt
 ```
 
+Note that the (optional) `@parcels.validate_kernel` decorator can be used to ensure that the kernel function is correctly defined and can be safely executed by parcels.
```
Comment (VeckoTheGecko) [inline docs/user_guide/getting_started/explanation_concepts.md]:
```suggestion
Note that the (optional) `@parcels.validate_kernel` decorator can be used to ensure that the kernel function is correctly defined and follows recommendations.
```
I think "can be safely executed by parcels." is a bit of a strong statement

**Review by erikvansebille — APPROVED** [review]

**Comment by VeckoTheGecko** [comment]
I realised that a bunch of our kernels in testing didn't satisfy the warnings, causing the test suite to fail.

I've adjusted the tests which made sense to change, and added `@ignore_kernel_warnings` to the tests where we know we're using this sort of API (`@ignore_kernel_warnings` is a new pytest filterwarnings mark that I added).

The only question I have is about periodic boundary conditions - does it make sense for users to to use `parcels.dx +=` for that instead of `particles.x = ...` for that @erikvansebille ? (I don't see the mention of periodic BCs in the docs - maybe we should address all this in another issue and PR)

**Comment by erikvansebille** [comment]
> The only question I have is about periodic boundary conditions - does it make sense for users to to use parcels.dx += for that instead of particles.x = ... for that @erikvansebille ? (I don't see the mention of periodic BCs in the docs - maybe we should address all this in another issue and PR)

Can you provide a bit more context? Where did you see this statement with `particles.x = ...`?

**Comment by VeckoTheGecko** [comment]
Looked into it again, and its all good.

I was looking at a kernel in the testing code that was setting `x` directly instead of doing `dx`. Using `dx` is fine (just as it was in v3:)

```python
def periodicBC(particle, fieldset, time):
    if particle.lon < fieldset.halo_west:
        particle_dlon += fieldset.halo_east - fieldset.halo_west
    elif particle.lon > fieldset.halo_east:
        particle_dlon -= fieldset.halo_east - fieldset.halo_west
```
